README.md
pytrain
=======

Python engine for controlling model railways with a Raspberry Pi


debian/changelog
pytrain (0.3.1-1) all; urgency=medium

  * Add on/of semantics for special switched.

 -- Wolfgang Glas <wolfgang.glas@iteg.at>  Sun, 15 Nov 2015 17:45:25 +0100

pytrain (0.3.0-1) all; urgency=medium

  * Integrate switch 12 introduce status polling.

 -- Wolfgang Glas <wolfgang.glas@iteg.at>  Sun, 15 Nov 2015 12:32:43 +0100

pytrain (0.2.1-1) all; urgency=medium

  * Integrate switch 11 for even more power control
  * Change HTTP port from 8000 to 8011

 -- Wolfgang Glas <wolfgang.glas@iteg.at>  Sun, 27 Sep 2015 12:07:49 +0200

pytrain (0.2.0-1) all; urgency=medium

  * Integrate switch 10 for power control

 -- Wolfgang Glas <wolfgang.glas@iteg.at>  Sat, 26 Sep 2015 13:22:51 +0200

pytrain (0.1.0-1) all; urgency=medium

  * Initial debianization.

 -- Wolfgang Glas <wglas85@gmail.com>  Wed, 17 Dec 2014 15:36:10 +0100


htdocs/js/pytrain.js
var pytrain = function() {

	// number of railroad switches
	var NSWITCHES = 12;
	
    var svg = document.getElementById("railtrack_baselayout");
	
    var svgDoc = (svg.contentWindow || svg.contentDocument);
    if (svgDoc.document) svgDoc = svgDoc.document;
    
    var switches = [];
    
    var switchHandler = function(switchState) {
		
    	console.log("Received new state:",switchState);
		    
    	for (var i=0;i<NSWITCHES;++i) {
    		switches[i].toggle(switchState[i]);
    	}
	}
    var failHandler = function(err) {
    	console.error(err);
    }

    var Deferred = function() {
    	    	
    	this.then = function(successHandler,failHandler) {
    		
        	this.successHandler = successHandler;
        	this.failHandler = failHandler;

    	}
    	
    	this.resolve = function() {
    		
    		this.successHandler && this.successHandler.apply(this,arguments);
    	}

    	this.reject = function() {
    		
    		this.failHandler && this.failHandler.apply(this,arguments);
    	}
    };
    
    
    var xhr = function(args) {
    	
    	var url = args.url;
    	var method = args.method || 'GET';
    	
    	var d = new Deferred();
    		
    	var xmlhttp = new XMLHttpRequest();
   	
    	xmlhttp.onreadystatechange=function() {
    		if (xmlhttp.readyState==4) {
    			
    			if (xmlhttp.status==200) {
                                if (!xmlhttp.responseType && args.responseType == 'json' && typeof xmlhttp.response == 'string') {
                                    d.resolve(JSON.parse(xmlhttp.response));
                                }
                                else {
    				    d.resolve(xmlhttp.response);
                                }
    			}
    			else {
    				d.reject("HTTP query ["+method+" "+url+"] failed with code "+xmlhttp.status);
    			}
    		}
    	};
    	xmlhttp.open(method,url,true);
    	if (args.body) {
    		xmlhttp.send(body);
    	}
    	else {
    		xmlhttp.send();
    	}
        if (args.responseType) {
                xmlhttp.responseType = args.responseType;
        }
    	return d;
    };
    
    
    var fetchInitialState = function() {
        xhr({
    	    url:"/toggleSwitch",
		    method:"GET",
	        responseType:"json"
        }).then(switchHandler);
    };
	
    var define_switch = function(lbl) {
	   
    	var rwswitch = {

    			link:  svgDoc.getElementById(lbl+"_link") || document.getElementById(lbl+"_link"),
    			node_0:  svgDoc.getElementById(lbl+"_0"),
    			node_1:  svgDoc.getElementById(lbl+"_1"),

    			state: 0,
    			
    			showState: function() {
    				this.node_0.style.display= this.state==0 ? "inline" : "none";
    				this.node_1.style.display= this.state!=0 ? "inline" : "none";
    			},
    			
    			toggle: function(newState) {
    				
    				if (this.state != newState) {
    					this.state = newState;
    					this.showState();
    				}
    			}
    	};
    	
        if (rwswitch.link) {

			var special = !svgDoc.getElementById(lbl+"_link"); 

			if (special) {
      			rwswitch.link.addEventListener("mousedown",function() {

    				xhr({
    					url:"/pressSwitch?id="+lbl,
    					method:"GET",
    					responseType:"json"
    				}).then(switchHandler);
					
					var releaseHandler = function() {
    					xhr({
    						url:"/releaseSwitch?id="+lbl,
    						method:"GET",
    						responseType:"json"
    					}).then(switchHandler);
						document.body.removeEventListener("mouseup",releaseHandler);
						document.body.removeEventListener("blur",releaseHandler);
					};
					
					document.body.addEventListener("mouseup",releaseHandler,false);
					document.body.addEventListener("blur",releaseHandler,false);
					
				},false);				
			}
			else {

      			rwswitch.link.addEventListener("mouseup",function() {
					
    				xhr({
    					url:"/toggleSwitch?id="+lbl,
    					method:"GET",
    					responseType:"json"
    				}).then(switchHandler);
				},false);
			}
    	}

    	rwswitch.showState();
    	return rwswitch;
    } 
    
    for (var i=0;i<NSWITCHES;++i) {
    	
    	var lbl = "layer_switch_" + (i+1);
    	   	
        switches.push(define_switch(lbl));
    }
    

    // fetch initial state on startup
	fetchInitialState();
	
	setInterval(fetchInitialState,1000)
};


src/pytrain/backend.py
'''
Created on 06.01.2015

@author: kinders
'''

try:
    from RPi import GPIO
except ImportError:
    from pytrain.GPIOStub import GPIO

from threading import Timer
import logging
 
log = logging.getLogger('pytrain.backend')
   
class Backend(object):
    '''
    classdocs
    '''

    def __init__(self):
        '''
        Constructor
        '''
        self.switchState = [ 0,0,0,0,0,0,0,0,0,0,1,1 ]
        self.coupling = [[1,9],[0,9],[3],[2],[5],[4],[-11],[8],[7],[0,1],[-7],None ]
        # GPIO PIN numbers of the switches
        # first is the 8-port breakout board, second the 4-port breakout board
        GPIO.setmode(GPIO.BOARD)
        self.gpioMap = [ 3,5,7,11,13,15,19,21,  10,12,16,18 ]
        for i in range(0,len(self.switchState)):
            pinid = self.gpioMap[i]
            log.info("Initializing PIN {}".format(pinid))
            GPIO.setup(pinid,GPIO.OUT)
            GPIO.output(pinid,self.switchState[i]);
        
    def reset(self):
        for i in range(0,len(self.switchState)):
            if self.switchState[i] != 0:
                pinid = self.gpioMap[i]
                log.info("Resetting PIN {}".format(pinid))
                GPIO.output(pinid,0);
        
    def setSwitch(self,i,switchState):
        if self.switchState[i] != switchState:
            self.toggleSwitch(i)
        
    def toggleSwitchInternal(self,i):
        pinid = self.gpioMap[i]
        self.switchState[i] = 1-self.switchState[i]
        log.info("switched number [{}], pinid [{}] new state is {}".format(i,pinid,self.switchState))
        GPIO.output(pinid,self.switchState[i]);

    def toggleSwitch(self,i):
        self.toggleSwitchInternal(i)
        cpl = self.coupling[i]
        if cpl != None:
            for j in cpl:
                if j < 0:
                    if self.switchState[i] != 1-self.switchState[-1-j]:
                        self.toggleSwitchInternal(-1-j)
                else:
                    if self.switchState[i] != self.switchState[j]:
                        self.toggleSwitchInternal(j)
        if i == 11:
            def timerfunc():
                if self.switchState[11]==0:
                    log.info("Resetting special purpose switch [11]")
                    self.toggleSwitchInternal(11)
            timer = Timer(10.0,timerfunc)
            timer.start()
            



src/pytrain/Main.py
import os
from werkzeug.wrappers import Request, Response, Headers
from werkzeug.wsgi import SharedDataMiddleware
import json
from pytrain.backend import Backend
import sys
import logging
import logging.handlers
import signal
import select

class PyTrainRequestHandler(object):

    def __init__(self, config):
        self.config = config
        self.backend = Backend()

    def dispatch_request(self, request):
        
        if request.path == '/':
            return Response("Found",status=302,headers=Headers({"Location": "/index.html"}))                   
        elif request.path == '/toggleSwitch' or request.path == '/releaseSwitch' or request.path == '/pressSwitch':
            
            if request.values.has_key('id'):
                swid = request.values['id']
                if swid and swid.startswith('layer_switch_'):
                    i = int(swid[13:])-1
                    if  request.path == '/releaseSwitch':
                        self.backend.setSwitch(i,1)
                    elif request.path == '/pressSwitch':
                        self.backend.setSwitch(i,0)
                    else:
                        self.backend.toggleSwitch(i)
                
            je = json.encoder.JSONEncoder()
            body = je.encode(self.backend.switchState)
            return Response(body,content_type="appplication/json")
                
        
        return Response('Not Found',status=404)

    def wsgi_app(self, environ, start_response):
        request = Request(environ)
        response = self.dispatch_request(request)
        return response(environ, start_response)

    def __call__(self, environ, start_response):
        return self.wsgi_app(environ, start_response)


def create_app(redis_host='localhost', redis_port=6379, with_static=True):
    
    app = PyTrainRequestHandler({
        'use_gpio':       False
    })
    if with_static:
        app.wsgi_app = SharedDataMiddleware(app.wsgi_app, {
            '/':  os.path.join(os.path.dirname(__file__), '../../htdocs')
        })
    return app

if __name__ == '__main__':

    if os.getuid() == 0:
# daemon-mode logging
        handler = logging.handlers.SysLogHandler(facility=logging.handlers.SysLogHandler.LOG_DAEMON,address='/dev/log')

        formatter = logging.Formatter('pytrain %(levelname)s %(message)s')
        handler.setFormatter(formatter)
    else:
# development-mode logging
        handler = logging.StreamHandler(open('/dev/stderr', 'w'))
        formatter = logging.Formatter( '%(asctime)s %(levelname)s %(message)s') 
        handler.setFormatter(formatter)
        
    root_logger = logging.getLogger()
    root_logger.addHandler(handler)
    root_logger.setLevel(logging.DEBUG)

    root_logger.debug("sys.path=[{}]".format(sys.path))
    # from werkzeug.serving import run_simple
    from werkzeug.serving import make_server
    app = create_app()
    
    server = make_server('0.0.0.0',8011,app,threaded=True)
    
    def signal_handler(sig,frame):
        root_logger.info("Shutting down server upon signal %d."%sig)
        app.backend.reset()
        os.close(server.fileno())
        
    signal.signal(signal.SIGINT,signal_handler)
    signal.signal(signal.SIGHUP,signal_handler)
    signal.signal(signal.SIGTERM,signal_handler)
    signal.signal(signal.SIGABRT,signal_handler)
    signal.signal(signal.SIGQUIT,signal_handler)
    
    try:
        server.serve_forever()
    except select.error:
        exit()
